targets/TARGET_Analog_Devices/TARGET_ADUCM4X50/TARGET_ADUCM4050/bsp/rng/adi_rng.c

```diff
@@ -0,0 +1,796 @@
+            during callbacks as the callback is executing context of the initiating interrupt and will
+            block lower-priority interrupts.  If extended application-level interrupt processing is
+            required, the application should schedule it for the main application loop and exit the
+            callback as soon as possible.\n
+
+
+    @sa    adi_rng_Open().
+*/
+ADI_RNG_RESULT adi_rng_RegisterCallback (
+                                         ADI_RNG_HANDLE  hDevice,
+                                         ADI_CALLBACK    cbFunc,
+                                         void           *pCBParam)
+{
+    ADI_RNG_DEV_TYPE *pDevice = (ADI_RNG_DEV_TYPE*)hDevice;
+
+#ifdef ADI_DEBUG
+   if (ADI_RNG_INVALID_HANDLE(pDevice)){
+        return ADI_RNG_BAD_DEV_HANDLE;
+    }
+
+    if (NULL == pDevice->pData) {
+        return ADI_RNG_NOT_INITIALIZED;
+    }
+#endif
+
+    /* save the callback info */
+    pDevice->pData->CBFunc   = cbFunc;
+    pDevice->pData->pCBParam = pCBParam;
+
+    if (NULL != cbFunc) {
+        /* enable RNG interrupts in NVIC */
+        NVIC_EnableIRQ(pDevice->pData->IRQn);
+    } else {
+        NVIC_DisableIRQ(pDevice->pData->IRQn);
+    }
+
+    return ADI_RNG_SUCCESS;
+}
+
+/*! \cond PRIVATE */
+/* RNG driver interrupt handler. Overrides weak default handler in startup file */
+void RNG_Int_Handler(void)
+{
+    ISR_PROLOG();
+    ADI_RNG_DEV_TYPE *pDevice = &gRNG_Device[0];
+    register uint16_t candidate;
+
+    /* if we have an initialized driver... */
+    if (NULL != pDevice->pData)
+    {
+        /* if we have a registered callback */
+        if (NULL != pDevice->pData->CBFunc)
+        {
+            ADI_INT_STATUS_ALLOC();
+
+            ADI_ENTER_CRITICAL_REGION();
+            /* read status register without other interrupts in between */
+            candidate = pDevice->pRNG->STAT;
+            ADI_EXIT_CRITICAL_REGION();
+
+            /* Only have bits in stat that are necessary */
+            candidate = candidate & (BITM_RNG_STAT_STUCK | BITM_RNG_STAT_RNRDY);
+
+            while (0u != candidate) {
+                uint32_t nEvent;
+
+                if (0u != (candidate & BITM_RNG_STAT_RNRDY)) {
+                    nEvent = ADI_RNG_EVENT_READY;
+                    candidate &= (uint16_t)~BITM_RNG_STAT_RNRDY;
+                } else if (0u != (candidate & BITM_RNG_STAT_STUCK)) {
+                    nEvent = ADI_RNG_EVENT_STUCK;
+                    candidate &= (uint16_t)~BITM_RNG_STAT_STUCK;
+                } else {
+                    break;
+                }
+
+                pDevice->pData->CBFunc (
+                                        pDevice->pData->pCBParam,
+                                        nEvent,
+                                        NULL
+                                        );
+            }
+
+            pDevice->pRNG->STAT = BITM_RNG_STAT_RNRDY | BITM_RNG_STAT_STUCK;
```

Comment:
Is there a reason why we don't use a guard here, but we do in the read from`pRNG->STAT` a few lines earlier?

Reply:
This write to the STAT register is to clear the RNRDY and STUCK bits which are set when a new random value is available. Writing to these bits does not trigger an interrupt and does not affect the TRNG hardware in generating the next value, hence disabling interrupt around it is not necessary.

Reply:
Thanks for the explanation! Two similar points:
1) `adi_rng_GetRdyStatus` also reads `pRNG->STAT`, but without a guard by `ADI_INT/EXIT_CRITICAL_REGION`. What's the reason for that difference?
2) To my understanding of the spec, if the callback called from `RNG_Int_Handler` reads the random data, `pRNG->STAT` will automatically be cleared and the RNG will accumulate new data. It then seems conceivable that by the time `RNG_Int_Handler` writes to `pRNG->STAT` for clearance, the TRNG has already updated `pRNG->STAT` again, and that we're therefore missing a signal from the TRNG. Is that a problem for the stuck signal? From reading the spec it's not clear in what way this condition is supposed to be handled. In particular, should one check for `BITM_RNG_STAT_STUCK` not being set (e.g. in `adi_rng_GetRdyStatus`) before reading the sampled data?

Reply:
Please see the answers below:

1.  I don’t think that interrupts need to be disabled in either case. In the handler there is no benefit to disabling interrupts when reading the STAT register. Copying the STAT register into a local variable doesn’t need protection, as it is not modified/used anywhere else at interrupt level.  
2. Yes, you are correct that there's a slight chance of the random number being updated between the time you write to the STAT and the time when you read the random number register. However if this does happen, the random number register would still contain a random number albeit a different one.

Reply:
Thanks for your answers.
1. Could you then unify the code please?
2. Could you address the question regarding the `STUCK` register?

Reply:
1. Since the enabling/disabling of the interrupt does not affect the correct operation of the RNG, I would rather not make changes to the BSP code now as it will diverge from the formal BSP release. I have already informed our BSP software engineer to make the changes for the next release, and I'll update the code when that becomes available. 
2. My understanding is that under normal operations the STUCK bit is never set, unless there is something wrong with the underlying hardware oscillator. In an event that this does happen, I guess trng_get_bytes would need to let the upper layer application know somehow. Not sure how trng_get_bytes could propagate this error up the HAL?

Reply:
1. Sounds good to me.
2. That's also my understanding, and I'd like to ensure that we don't return anything from `trng_get_bytes` if `STUCK` is set. Intuitively, `STUCK` and `RNRDY` should be mutually exclusive, so checking `RNRDY` as is currently done in `trng_get_bytes` _should_ be enough, but I didn't find any indication of that in the spec (?). A way to deal with this uncertainty would be to also call `adi_rng_GetStuckStatus` in `trng_get_bytes` and only proceed to fetching data if it's unset. Otherwise, the function might return `-1` to signal the error.

Reply:
Added a check for the STUCK bit in the latest commit.